### DISCUSSION TITLE
Is it possible to have Objects as Subject Sets?

### DISCUSSION THREAD

**Discussion by thestephenstanton** [ory/keto#767] [Q&A] [answered]
Disclaimer: I'm super new to Keto, like a few days. So ELI5 please.

It is best described with an example:

## Requirements
1. Any file in a directory can be read by any reader of a directory.
2. Any file in a directory can be written to by any writer of a directory.

## Initial setup
this part will be the same for both solutions
```
dir1#reader@user1
dir1#writer@user2
```

## Current Solution

readers can read
```
file1#read@(dir1#reader)
file2#read@(dir1#reader)
file3#read@(dir1#reader)
```

writers can write
```
file1#writer@(dir1#writer)
file2#writer@(dir1#writer)
file3#writer@(dir1#writer)
```

number of api calls:
```
# of files * # of relations
```

## Proposed Solution

files are children of a directory
```
dir1#child@file1
dir1#child@file2
dir1#child@file3
```

readers can read
```
(dir1#child)#read@(dir1#reader)
```

writers can write
```
(dir1#child)#write@(dir1#writer)
```

number of api calls:
```
# of files + # of relations
```

## Final Words
Mainly wondering if this can be done or if there are limitations with the graph algorithms that don't make this possible. Or if there is a similar way to achieve this.

I'm trying to avoid having to make so many api calls when there could be many relations to a given object. For instance say I had reader, writer, sharer, deleter, I'd have to make 4 relation tuples each time I created a file.

I think it is also worth mentioning, I am trying to preserve the hierarchy via Keto which maybe is a smell. Meaning I want users to have "roles" at a directory level that cascade into the child objects.

**Comment by zepatrik** [comment, marked as answer]
FYI, just converted to discussion as I think it is rather a discussion instead of issue :wink: 

The proposed solution is not going to be very performant in the end as it enables to build very complex graphs that have many paths. Instead, the idea would be to have the `child` or a `parent` relation (depending on your needs) as you proposed, but then use subject set rewrites to define that you have "read when you have read to the parent" (#263 and other issues). This basically means that you only need `# of files` API calls. Note that unfortunately rewrites are currently not yet implemented, and you need to use the workaround you proposed in very beginning.

**Reply by vinckr** [reply]
Hey @thestephenstanton ,
Just checking in if your question was answered or if there was still something unclear.
I will mark this as answered for now, but feel free to come back to the discussion any time.

Thanks a ton! 🐝
